Source/WTF/wtf/ThreadSafeWeakPtr.h

```diff
@@ -0,0 +1,202 @@
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/Lock.h>
+#include <wtf/MainThread.h>
+#include <wtf/RefPtr.h>
+
+namespace WTF {
+
+template<typename> class ThreadSafeWeakPtr;
+template<typename, DestructionThread> class ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr;
+
+template<typename T>
+class ThreadSafeWeakPtrControlBlock {
+    WTF_MAKE_NONCOPYABLE(ThreadSafeWeakPtrControlBlock);
+    WTF_MAKE_STRUCT_FAST_ALLOCATED;
+public:
+    void ref() const
+    {
+        Locker locker { m_lock };
+        m_weakReferenceCount++;
+    }
+
+    void deref() const
+    {
+        bool shouldDeleteControlBlock { false };
+        {
+            Locker locker { m_lock };
+            RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_weakReferenceCount);
+            if (!--m_weakReferenceCount && !m_strongReferenceCount)
+                shouldDeleteControlBlock = true;
+        }
+        if (shouldDeleteControlBlock)
+            delete this;
+    }
+
+    void strongRef() const
+    {
+        Locker locker { m_lock };
+        RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_object);
+        m_strongReferenceCount++;
+    }
+
+    template<DestructionThread destructionThread>
+    void strongDeref() const
+    {
+        bool shouldDeleteControlBlock { false };
+        {
+            Locker locker { m_lock };
+            RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_object);
+            if (LIKELY(--m_strongReferenceCount))
+                return;
+
+            auto deleteObject = [object = std::exchange(m_object, nullptr)] {
+                delete static_cast<const T*>(object);
+            };
+            switch (destructionThread) {
+            case DestructionThread::Any:
+                deleteObject();
+                break;
+            case DestructionThread::Main:
+                ensureOnMainThread(WTFMove(deleteObject));
+                break;
+            case DestructionThread::MainRunLoop:
+                ensureOnMainRunLoop(WTFMove(deleteObject));
+                break;
+            }
+
+            if (!m_weakReferenceCount)
+                shouldDeleteControlBlock = true;
+        }
+        if (shouldDeleteControlBlock)
+            delete this;
+    }
+
+    RefPtr<T> makeStrongReferenceIfPossible() const
+    {
+        Locker locker { m_lock };
+        if (m_object) {
+            // Calling the RefPtr constructor would call strongRef() and deadlock.
+            m_strongReferenceCount++;
+            return adoptRef(m_object);
+        }
+        return nullptr;
+    }
+
+    size_t refCount() const
```

Comment:
Do we need this function for anything? It seems somewhat of limited value for a thread safe refcounted object, since the answer can just change at any time.

The exception I guess is for the `hasOneRef()` case, where if it's true, you can then reason about being the only owner (and it can't change behind your back). Can we just have that function instead?

Reply:
refCount and hasOneRef can both change from another thread if weak pointers exist.  I think we should remove both.

Reply:
I think `hasOneRef` should only be true if there were no weak refs, but agree that it's just better not to offer it until we have a compelling need.